torch/optim/lbfgs.py

```diff
@@ -0,0 +1,243 @@
+import torch
+from functools import reduce
+from .optimizer import Optimizer
+
+
+class LBFGS(Optimizer):
+    """Implements L-BFGS algorithm.
+
+    .. warning::
+        This optimizer doesn't support per-parameter options and parameter
+        groups (there can be only one).
+
+    .. warning::
+        Right now all parameters have to be on a single device. This will be
+        improved in the future.
+
+    .. note::
+        This is a very memory intensive optimizer (it requires additional
+        ``param_bytes * history_size`` bytes). If it doesn't fit in memory
+        try reducing the history size, or use a different algorithm.
+
+    Arguments:
+        lr (float): learning rate (default: 1)
+        max_iter (int): maximal number of iterations per optimization step
+            (default: 20)
+        max_eval (int): maximal number of function evaluations per optimization
+            step (default: max_iter * 1.25).
+        tolerance_grad (float): termination tolerance on first order optimality
+            (default: 1e-5).
+        tolerance_change (float): termination tolerance on function value/parameter
+            changes (default: 1e-9).
+        history_size (int): update history size (default: 100).
+    """
+
+    def __init__(self, params, lr=1, max_iter=20, max_eval=None,
+            tolerance_grad=1e-5, tolerance_change=1e-9, history_size=100,
+            line_search_fn=None):
+        if max_eval is None:
+            max_eval = max_iter * 5 // 4
+        defaults = dict(lr=lr, max_iter=max_iter, max_eval=max_eval,
+                        tolerance_grad=tolerance_grad, tolerance_change=tolerance_change,
+                        history_size=history_size, line_search_fn=line_search_fn)
+        super(LBFGS, self).__init__(params, defaults)
+
+        if len(self.param_groups) != 1:
+            raise ValueError("LBFGS doesn't support per-parameter options "
+                    "(parameter groups)")
+
+        self._params = self.param_groups[0]['params']
+        self._numel_cache = None
+
+    def _numel(self):
+        if self._numel_cache is None:
+            self._numel_cache = reduce(lambda total, p: total + p.numel(), self._params, 0)
+        return self._numel_cache
+
+    def _gather_flat_grad(self):
+        return torch.cat(
+                tuple(param.grad.data.view(-1) for param in self._params), 0)
+
+    def _add_grad(self, step_size, update):
+        offset = 0
+        for p in self._params:
+            numel = p.numel()
+            p.data.add_(step_size, update[offset:offset+numel])
+            offset += numel
+        assert offset == self._numel()
+
+    def step(self, closure):
+        """Performs a single optimization step.
+
+        Arguments:
+            closure (callable): A closure that reevaluates the model
+                and returns the loss.
+        """
+        # TODO: backward after closure?
```

Comment:
I didn't remember if it's the responsibility of a closure, or the optimizer to call `.backward()`. I'll just remove it.